compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3471,43 +3475,82 @@ TR::Register *OMR::Power::TreeEvaluator::vmulEvaluator(TR::Node *node, TR::CodeG
      }
    }
 
-TR::Register *OMR::Power::TreeEvaluator::vmulInt32Helper(TR::Node *node, TR::CodeGenerator *cg)
+TR::Register *OMR::Power::TreeEvaluator::vmulInt8Helper(TR::Node *node, TR::CodeGenerator *cg)
    {
    TR::Node *firstChild;
    TR::Node *secondChild;
    TR::Register *lhsReg, *rhsReg;
-   TR::Register *resReg;
-   TR::Register *tempA;
-   TR::Register *tempB;
-   TR::Register *tempC;
+   TR::Register *productReg;
+   TR::Register *temp;
+   TR::Register *shiftReg;
 
    firstChild = node->getFirstChild();
    secondChild = node->getSecondChild();
-   lhsReg = NULL;
-   rhsReg = NULL;
 
    lhsReg = cg->evaluate(firstChild);
    rhsReg = cg->evaluate(secondChild);
 
-   resReg = cg->allocateRegister(TR_VRF);
-   tempA = cg->allocateRegister(TR_VRF);
-   tempB = cg->allocateRegister(TR_VRF);
-   tempC = cg->allocateRegister(TR_VRF);
-   node->setRegister(resReg);
-   generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisw, node, tempA, -16);
-   generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisw, node, tempB, 0);
-   generateTrg1Src2Instruction(cg, TR::InstOpCode::vmulouh, node, tempC, lhsReg, rhsReg);
-   generateTrg1Src2Instruction(cg, TR::InstOpCode::vrlw, node, resReg, rhsReg, tempA);
-   generateTrg1Src3Instruction(cg, TR::InstOpCode::vmsumuhm, node, tempB, lhsReg, resReg, tempB);
-   generateTrg1Src2Instruction(cg, TR::InstOpCode::vslw, node, tempA, tempB, tempA);
-   generateTrg1Src2Instruction(cg, TR::InstOpCode::vadduhm, node, resReg, tempA, tempC);
-
-   cg->stopUsingRegister(tempA);
-   cg->stopUsingRegister(tempB);
-   cg->stopUsingRegister(tempC);
+   productReg = cg->allocateRegister(TR_VRF);
+   temp = cg->allocateRegister(TR_VRF);
+   shiftReg = cg->allocateRegister(TR_VRF);
+
+   node->setRegister(productReg);
+
+   //set shift amount to 1 byte
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisb, node, shiftReg, 8);
+
+   //multiply even bytes and shift to correct location
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vmuleub, node, productReg, lhsReg, rhsReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vslh, node, productReg, productReg, shiftReg);
```

Comment:
Here's a quick illustration of what happens in gdb:

lhsReg = {24, 26, 28, 30, 16, 18, 20, 22, 8, 10, 12, 14, 0, 2, 4, 6}
rhsReg = {12, 13, 14, 15, 8, 9, 10, 11, 4, 5, 6, 7, 0, 1, 2, 3}

productReg = vmuleub(lhsReg,rhsReg) = {82, 1, -62, 1, -94, 0, -14, 0, 50, 0, 98, 0, 2, 0, 18, 0}
productReg = vslh(productReg, 8) = {0, 82, 0, -62, 0, -94, 0, -14, 0, 50, 0, 98, 0, 2, 0, 18}

So in this case, the even bytes are 26, 30, 18...2, 6 in lhsReg and 13, 15, 9...1, 3 in rhsReg. Shifting their products by 8 bytes puts them in the correct place (i.e.: in the even spaces).

Reply:
Ok, so according to you it's (I just formatted the example above slightly) 
```lhsReg  = {24, 26, 28, 30, 16, 18,  20, 22,  8, 10, 12, 14, 0, 2,  4,  6}
rhsReg = {12, 13, 14, 15,  8,  9,  10, 11,  4,  5,  6,  7, 0, 1,  2,  3}
vmuleub= {82,  1,-62,  1,-94,  0, -14,  0, 50,  0, 98,  0, 2, 0, 18,  0}

```
But I think according to the documentation it's : 
`vmuleub= {                  ...             0, 50,  0, 98, 0, 2,  0, 18}`

In the debugger, are you looking at a vector register or value in memory? There might be an LE issue there.

ISA is saying that each multiplication is placed into a half word. So if result of multiplying elements 0 is 18, resulting half word 0 should have (0, 18) in it, not (18, 0).

Reply:
I think the reason it's (18, 0) is because it's LE. 18 is the lower half of the product, and 0 is the upper half. The values above are from printing out the contents of the vector registers, not memory, if that makes any difference.

Reply:
It does for sure. It just occurred to me what could go wrong. Could you please paste the whole sequence of instructions including the ones that load the source registers and store the result?

Reply:
Please also print the arrays you are adding in Java (from 0 to array length-1)

Reply:
Here is the full sequence of instructions starting from when the vectors are loaded into lhsReg and rhsReg to when productReg (i.e.: the final answer) is stored to memory: 
```
   0x7fffc448f058:	li      r3,128
   0x7fffc448f05c:	lxvw4x  vs33,r14,r3
   0x7fffc448f060:	li      r3,112
   0x7fffc448f064:	lxvw4x  vs35,r14,r3
   0x7fffc448f068:	vspltisb v2,8
   0x7fffc448f06c:	vmuleub v0,v1,v3
   0x7fffc448f070:	vslh    v0,v0,v2
   0x7fffc448f074:	vmuloub v1,v1,v3
   0x7fffc448f078:	vslh    v1,v1,v2
   0x7fffc448f07c:	vsrh    v1,v1,v2
   0x7fffc448f080:	vaddubm v0,v0,v1
   0x7fffc448f084:	extsw   r3,r27
   0x7fffc448f088:	addi    r3,r3,16
=> 0x7fffc448f08c:	stxvw4x vs32,r3,r21
```
And here are the original java arrays that are used to form the vectors that are then multiplied together: 

a[] = {0, 2, 4, 6, 8, 10, 12, 14, 16, ...}
b[] = {0, 1, 2, 3, 4, 5, 6, 7, 8, ...}

Reply:
Thanks! lxvw4x explains the incorrect order of bytes on LE platform. We definitely need to fix that. However, I realized that I was counting bytes incorrectly (confused which ones are even and which ones are odd in the register). I agree that even ones need to be shifted since the left most byte is 0.

Reply:
The use of `lxvw4x` doesn't cause any problems here, since we're working in bytes and halfwords, and `stxvw4x` is used to store the result, but it would probably make more sense in a situation like this to use the byte versions of these instructions (`lxvb16x` and `stxvb16x`).

That being said, this isn't really within the scope of this particular issue, and it's not causing any immediate problems with this one, so I think we can save that for another PR.